src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,1013 @@
+                    {
+                        switch (decisionTree.Kind)
+                        {
+                            case DecisionTree.DecisionKind.ByValue:
+                                return 0; // a value pattern is always considered incomplete (even bool true and false)
+                            case DecisionTree.DecisionKind.ByType:
+                                {
+                                    var byType = (DecisionTree.ByType)decisionTree;
+                                    if (inputCouldBeNull &&
+                                        (byType.WhenNull == null || CheckSubsumed(pattern, byType.WhenNull, inputCouldBeNull) == 0) &&
+                                        (byType.Default == null || CheckSubsumed(pattern, byType.Default, inputCouldBeNull) == 0))
+                                        return 0; // new pattern catches null if not caught by existing WhenNull or Default
+                                    inputCouldBeNull = false;
+                                    foreach (var td in byType.TypeAndDecision)
+                                    {
+                                        var type = td.Key;
+                                        var decision = td.Value;
+                                        if (_conversions.ExpressionOfTypeMatchesPatternType(decisionTree.Type, type) == true)
+                                        {
+                                            var error = CheckSubsumed(pattern, decision, inputCouldBeNull);
+                                            if (error != 0) return error;
+                                        }
+                                    }
+                                    if (byType.Default != null)
+                                    {
+                                        return CheckSubsumed(pattern, byType.Default, inputCouldBeNull);
+                                    }
+                                    return 0;
+                                }
+                            case DecisionTree.DecisionKind.Guarded:
+                                {
+                                    var guarded = (DecisionTree.Guarded)decisionTree;
+                                    return (guarded.Guard == null || guarded.Guard.ConstantValue == ConstantValue.True) ? ErrorCode.ERR_PatternIsSubsumed :
+                                        guarded.Default != null ? CheckSubsumed(pattern, guarded.Default, inputCouldBeNull) : 0;
+                                }
+                            default:
+                                throw ExceptionUtilities.UnexpectedValue(decisionTree.Kind);
+                        }
+                    }
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(pattern.Kind);
+            }
+        }
+
+        internal delegate DecisionTree DecisionMaker(
+            BoundExpression expression,
+            TypeSymbol type);
```

Comment:
Will make it private in a follow-up PR.

---
In reply to: [68627671](https://github.com/dotnet/roslyn/pull/11485#discussion_r68627671) [](ancestors = 68627671)